src/main/java/com/zendesk/maxwell/producer/MaxwellKafkaProducer.java

```diff
@@ -231,18 +240,36 @@ public void sendAsync(RowMap r, AbstractAsyncProducer.CallbackCompleter cc) thro
 
 		KafkaCallback callback = new KafkaCallback(cc, r.getNextPosition(), r.getRowIdentity(), value,
 				this.succeededMessageCount, this.failedMessageCount, this.succeededMessageMeter, this.failedMessageMeter,
-				this.deadLetterTopic, this.context, this);
+				this.topic, this.deadLetterTopic, this.context, this);
 
 		sendAsync(record, callback);
 	}
 
-	public void sendFallbackAsync(String topic, RowIdentity fallbackRecord, KafkaCallback callback, RecordMetadata md, Exception reason) {
-		try {
-			ProducerRecord<String, String> record = makeFallbackRecord(topic, fallbackRecord, reason);
-			sendAsync(record, callback);
-		} catch (Exception fallbackEx) {
-			callback.onCompletion(md, fallbackEx);
+	private ExecutorService getBackgroundExecutor() {
+		if (this.backgroundExecutor == null) {
+			this.backgroundExecutor = Executors.newSingleThreadExecutor();
 		}
+		return this.backgroundExecutor;
+	}
+
+	public void sendFallbackAsync(String topic, RowIdentity fallbackRecord, KafkaCallback callback, RecordMetadata md, Exception reason) {
+		// Note: this code may be executed from two different threads:
+		//  - client rejection: invoked synchronously from the  MaxwellKafkaProducerWorker thread (in sendAsync())
+		//  - broker rejection: invoked asynchronously from kafka's `kafka-producer-network-thread`
+		// We can't call `sendAsync` on kafka's own network thread, it causes a deadlock. So we
+		// invoke the send via an Executor with its own thread.
+		getBackgroundExecutor().submit(new Runnable() {
```

Comment:
Can there be multiple concurrent calls to `sendFallbackAsync`?

Reply:
Good point. Probably? It can be called from 2 threads as outlined, and I don't think there's any synchronization between them :/

Are you worried about `getBackgroundExecutor` making more than one pool due to races? I considered just using `ForkJoinPool.commonPool()`, which is easier but slightly cheekier if you're embedding maxwell in an app which uses that for its own purposes.

Reply:
Was worried that the backgroundExecutor wasn't defined as `volatile` so could potentially break if two concurrent calls.  

More than 1 pool probably isn't too bad, I assume it would get GCed pretty quickly without a reference to it.

Reply:
believe that we can use lambas here for a tad of code cleanliness.